sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/FixedLengthRowBasedKeyValueBatch.java

```diff
@@ -165,10 +165,16 @@ private void freeCurrentPage() {
   protected FixedLengthRowBasedKeyValueBatch(StructType keySchema, StructType valueSchema,
                                              int maxRows, TaskMemoryManager manager) {
     super(keySchema, valueSchema, maxRows, manager);
-    klen = keySchema.defaultSize()
-            + UnsafeRow.calculateBitSetWidthInBytes(keySchema.length());
-    vlen = valueSchema.defaultSize()
-            + UnsafeRow.calculateBitSetWidthInBytes(valueSchema.length());
+    int keySize = 0;
+    int valueSize = 0;
+    for (String name : keySchema.fieldNames()) {
+      keySize += (keySchema.apply(name).dataType().defaultSize() + 7) / 8 * 8;
```

Comment:
Let's add a small comment about this implicit ceiling logic and the reason why `schema.defaultSize()` doesn't work.